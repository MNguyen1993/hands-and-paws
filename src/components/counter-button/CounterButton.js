import React, { Component } from 'react';

class CounterButton extends Component {
	constructor() {
		super();
		this.state = {
			count: 0,
		};
	}

	shouldComponentUpdate(nextProps, nextState) {
		if (this.state.count !== nextState.count) {
			return true;
		}
		return false;
	}

	updateCount = () => {
		// state updates are not synchronous, this may cause unwanted behavior
		// this.setState({ count: this.state.count + 1 });

		// Recommended way of updating state over the above when dependent on current state
		this.setState(state => ({ count: state.count + 1 }));
	};

	render() {
		return (
			<button
				className="b--none bg-light-blue br2 pa2 pointer"
				color={this.props.color}
				onClick={this.updateCount}
			>
				Adopted Pets: {this.state.count}
			</button>
		);
	}
}

export default CounterButton;
