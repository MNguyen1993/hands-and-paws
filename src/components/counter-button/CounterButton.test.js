import { shallow } from 'enzyme';
import React from 'react';
import CounterButton from './CounterButton';

describe('CounterButton Component', () => {
	const mockColor = 'blue';
	const mockNextProps = {};
	const mockNextState = {
		count: 1,
	};
	const mockNextState2 = {
		count: 0,
	};
	const wrapper = shallow(<CounterButton color={mockColor} />);
	const componentInstance = wrapper.instance();

	it('CounterButton component renders correctly', () => {
		expect(wrapper).toMatchSnapshot();
	});

	it('props are passed through correctly', () => {
		expect(componentInstance.props.color).toEqual('blue');
	});

	it('shouldComponentUpdate() returns true when the count changes', () => {
		expect(
			componentInstance.shouldComponentUpdate(mockNextProps, mockNextState)
		).toBe(true);
	});

	it('shouldComponentUpdate() returns false when the count is the same', () => {
		expect(
			componentInstance.shouldComponentUpdate(mockNextProps, mockNextState2)
		).toBe(false);
	});

	it('updateCount increments the count', () => {
		componentInstance.updateCount();
		expect(componentInstance.state.count).toBe(1);
	});

	it('button click increments the count', () => {
		wrapper.find('button').simulate('click');
		expect(wrapper.state()).toEqual({ count: 2 });
	});
});
