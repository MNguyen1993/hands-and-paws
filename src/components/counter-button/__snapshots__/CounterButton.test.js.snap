// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CounterButton Component CounterButton component renders correctly 1`] = `
<button
  className="b--none bg-light-blue br2 pa2 pointer"
  color="blue"
  onClick={[Function]}
>
  Adopted Pets: 
  0
</button>
`;
